plugins/osm-quickedit/public/main.js

```diff
@@ -1,26 +1,146 @@
+					if($('.sharePopup.top').is(':visible')){ 
+						// it is very unlikely, but possible that the sharePopup is open
+						// in which case we need to click the checkbox not the button
+						$('.shareButton div.checkbox input').click()
+					} else {
+						// otherwise we can just click the shareButton and it will turn on sharing
+						$('.shareButton').click()
+					}
+				}
+			}
+			
+			if (editor === 'ideditor'){
+				var editUrl = "https://www.openstreetmap.org/edit?editor=id#map=" +
+					options.map.getZoom() + "/" + 
+					options.map.getCenter().lat + "/" + 
+					options.map.getCenter().lng +
+					"&background=custom:" + tileUrl;	
+				window.open(editUrl);
+			} 
+			else if (editor === 'josm'){
+				
+				function formatUrlParams_(params){
+		      return "?" + Object
+		              .keys(params)
+		              .map(function (key){
+		                return key + "=" + params[key]
+		              })
+		              .join("&")
+		    }
+				
+				function sendJOSMCmd(endpoint, params){
+						var url = endpoint + formatUrlParams_(params),
+								loaded,
+								iframe;
+
+						return new Promise(function(resolve, reject){
+								// Figure out when we can next run a command
+								var wait = Math.max(josmLastCommand + JOSM_COMMAND_TIMEOUT - Date.now(), 0);
 
-		return React.createElement("button", {
-				type: "button",
-				className: "btn btn-sm btn-secondary",
-				onClick: function(){
-					var mapLocation = options.map.getZoom() + "/" + 
-									  options.map.getCenter().lat + "/" + 
-									  options.map.getCenter().lng;
-					window.location.href = "https://www.openstreetmap.org/edit?editor=id#map=" + mapLocation + 
-											"&background=custom:" + url;
+								// This remembers when we are going to run THIS command, and adds the timeout 
+								// (yes, it is double-counted - this seems to be more reliable).
+								josmLastCommand = Date.now() + wait + JOSM_COMMAND_TIMEOUT;
+
+								setTimeout(function(){
+										iframe = document.createElement('iframe');
+										iframe.style.display = 'none';
+										iframe.addEventListener('load', function(){
+												if(loaded === undefined){
+														loaded = true;
+														resolve();
+														iframe.parentElement.removeChild(iframe);
+												}
+										});
+										iframe.setAttribute('src', url);
+										document.body.appendChild(iframe);
+								}, wait);
+
+								setTimeout(function(){
+										if(loaded === undefined){
+												loaded = false;
+												reject();
+												iframe.parentElement.removeChild(iframe);
+										}
+								}, wait + JOSM_COMMAND_TIMEOUT);
+						});
+				}
+				
+				function prepareJOSM(){
+					var imageryParams = {
+		        title: tile.meta.name,
+		        type: "tms",
+		        url: encodeURIComponent("tms[22]:" + tileUrl)
+	        };
+					sendJOSMCmd('http://127.0.0.1:8111/imagery', imageryParams)				
```

Comment:
The JOSM button doesn't seem to work, I suspect because of this hard-coded address.

Reply:
Rest looks great!

Reply:
Do you have JOSM open on your computer? Nothing will happen if it's not. That's the behavior for both https://www.openstreetmap.org/ and https://tasks.hotosm.org/

Reply:
I thought those websites used to warn you if there wasn't a response and would show a prompt asking you if JOSM was open. But I don't see that happening now.

Reply:
Ahh, no I don't, what's why. Never mind then! Looks good!